### ISSUE TITLE
`spaceSplit(tgeompoint, integer)` returns different results with bitmatrix on vs off.

### ISSUE THREAD

**Issue by LauJosefsen** [MobilityDB/MobilityDB#313]
**Describe the bug**
Hi again :)

The parameter `bitmatrix` of the `spaceSplit` procedure seem to impact more than just performance. I don't know if this is intentional, but if it is, it should probably be declared in the documentation for the parameter.

The issue we faced is that some segments are left out if `bitmatrix:=true`. See steps to reproduce below.

**To Reproduce**
```sql
SELECT spaceSplit(
        tgeompoint 'SRID=3034;[POINT(4065298.6591165178 3154193.030669743)@2021-01-01 19:23:49+00, POINT(4064645.4035954443 3155752.896026873)@2021-01-01 19:28:09+00]',
        5000,
        bitmatrix := false
    );
```
Produces the three results, as expected.

```sql
SELECT spaceSplit(
        tgeompoint 'SRID=3034;[POINT(4065298.6591165178 3154193.030669743)@2021-01-01 19:23:49+00, POINT(4064645.4035954443 3155752.896026873)@2021-01-01 19:28:09+00]',
        5000,
        bitmatrix := true
    );
```
Only produces 2 results, i.e. an expected row is left out. The only change is the bitmatrix parameter.

**Expected behavior**
I would expect a parameter which is documented as an performance improvement to not change the output.

**Screenshots**
The desired output of 3 connected linestrings:
![image](https://user-images.githubusercontent.com/44977457/231699531-2fa00f71-363c-41b5-a1c7-f3dc75abd57c.png)

The undesired output, with `bitmatrix:=true` resulting in 2 unconnected linestrings:
![image](https://user-images.githubusercontent.com/44977457/231699475-964d0d5b-7441-4b61-b9e2-1e63d4f59e71.png)

**Specifications (please complete the following information):**
Use the commands:
```SQL
SELECT version();
-- PostgreSQL 15.2 (Ubuntu 15.2-1.pgdg22.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, 64-bit
SELECT postgis_full_version();
-- POSTGIS="3.3.2 4975da8" [EXTENSION] PGSQL="150" GEOS="3.10.2-CAPI-1.16.0" PROJ="8.2.1" GDAL="GDAL 3.4.1, released 2021/12/27" LIBXML="2.9.13" LIBJSON="0.15" LIBPROTOBUF="1.3.3" WAGYU="0.5.0 (Internal)" RASTER
SELECT mobilitydb_full_version();
-- MobilityDB 1.1.0, PostgreSQL 15.2 (Ubuntu 15.2-1.pgdg22.04+1), PostGIS 3.3.2, GEOS 3.10.2-CAPI-1.16.0, PROJ 8.2.1
```
Issue confirmed on build of latest commit on develop branch.

**Comment by estebanzimanyi** [comment]
Many thanks for reporting this bug and for your targeted example that helped us to understand the problem. After analysis I didn't find an easy fix for it :-(
The Bresenham line algorithm that is used for constructing the bitmap implicitly assumes that the points for determining the selected cells are in the center of the cell which is obviously not the case, this is a mistake in our design. In your example the cells are (1,0) and (0,1) which is a diagonal and then it misses the cell with the red segment in your figure.
How much gain is obtained in your use case when using the bitmap ? Is it foreseeable not to use it ? Any other suggestion for constructing the bitmap in a fast way ? A possibility that I see is to do not use the Bresenham algorithm and set all the cells in the bounding box defined by the selected cells of each segment. In your example, from the cells of the segment (1,0) and (0,1) it will select all 4 cells (0,0), (0,1), (1,0) and (1,1). However, the result of this method is that it will add many false positives and then we lose the advantage of using the bitmap.
Alas, in my answer there are many questions and not any answer :-( If you want, we can continue this conversation offline, you can easily find my email address and we can brainstorm for finding a solution to this problem.

**Comment by LauJosefsen** [comment]
Hi again.

Our use case is splitting trajectories of vessels in a day to day basis. We split them into a 5 km grid as part of an ETL process, where future queries then process the split trajectories further. On the limited data we have, it seems that running without bitmatrix is fine for our use case. It added ~20 seconds to a ~30 second query, but as it is ETL queries it doesn't really matter in the big picture. I just thought I'd raise the issue in here, as it was a bit unexpected, and we spent a fair amount of time trying to find the cause to a data artifact in our output data.

As for proposals, I must admit I am fairly new to the Moving Object world. The only thing I can think if is maybe using some form of thick line approach, with a line thickness of the space split size? It would probably still produce some false positives though? I found this if interested: http://homepages.enterprise.net/murphy/thickline/index.html

But yes, I'm afraid that using a bounding box for trajectories will involve too many false positives, as bounding boxes of trajectories, at least in our experience, grow in spatial area very quickly as trajectories become longer. This is the main reason we split the trajectories into the 5km spatial split to begin with, as the large bounding boxes is a nightmare in r-trees as well.

I don't know how much help I can provide in discussing a solution for the problem, but I will gladly try my best if you would like to discuss something over email or on this issue. :)
